### PR TITLE
Codechange: Store and pass Colours in Colours type.

### PR DESCRIPTION
**Pull request by PeterN**
## Motivation / Problem

Values from the Colours enum are often stored as a byte and passed as an integer, and may require casts to be used as a Colours again. This is just messy.

## Description

Store and pass colours in Colours type instead. This reduces casts, some magic numbers, and introduces a bit of type-safety, as well as reducing ambiguity.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

**Review by LordAro — COMMENTED** [review]
Casting is normally `(Colours)foo` rather than `Colours(foo)`

File: src/newgrf.cpp [hunk 1]
```diff
@@ -2405,10 +2405,10 @@ static ChangeInfoResult TownHouseChangeInfo(uint hid, int numinfo, int prop, Byt
 					housespec->grf_prop.local_id = hid + i;
 					housespec->grf_prop.subst_id = subs_id;
 					housespec->grf_prop.grffile = _cur.grffile;
-					housespec->random_colour[0] = 0x04;  // those 4 random colours are the base colour
-					housespec->random_colour[1] = 0x08;  // for all new houses
-					housespec->random_colour[2] = 0x0C;  // they stand for red, blue, orange and green
-					housespec->random_colour[3] = 0x06;
+					housespec->random_colour[0] = COLOUR_RED; // these 4 random colours are the base colour
```
Comment (2TallTyler) [inline src/newgrf.cpp]: Not your doing, but an inline comment wrapping over two lines and trying to describe four lines is very creative!

Maybe fix that. 😉

File: src/saveload/oldloader_sl.cpp [hunk 2]
```diff
@@ -465,10 +465,10 @@ static void FixTTOCompanies()
 	}
 }
 
-static inline byte RemapTTOColour(byte tto)
+static Colours RemapTTOColour(Colours tto)
```
Comment (2TallTyler) [inline src/saveload/oldloader_sl.cpp]: Lost its `inline` - intentional?

File: src/company_gui.cpp [hunk 3]
```diff
@@ -582,7 +582,7 @@ static const LiveryClass _livery_class[LS_END] = {
 template <SpriteID TSprite = SPR_SQUARE>
 class DropDownListColourItem : public DropDownIcon<DropDownString<DropDownListItem>> {
 public:
-	DropDownListColourItem(int colour, bool masked) : DropDownIcon<DropDownString<DropDownListItem>>(TSprite, PALETTE_RECOLOUR_START + (colour % COLOUR_END), colour < COLOUR_END ? _colour_dropdown[colour] : STR_COLOUR_DEFAULT, colour, masked)
+	DropDownListColourItem(int colour, bool masked) : DropDownIcon<DropDownString<DropDownListItem>>(TSprite, GENERAL_SPRITE_COLOUR(colour % COLOUR_END), colour < COLOUR_END ? _colour_dropdown[colour] : STR_COLOUR_DEFAULT, colour, masked)
```
Comment (2TallTyler) [inline src/company_gui.cpp:585]: `GENERAL_SPRITE_COLOUR` shows up several times but doesn't appear to originate in this PR. Is it a macro? (It looks like a constant but takes arguments like a function or constructor...)

File: src/newgrf.cpp [hunk 4]
```diff
@@ -2502,7 +2502,7 @@ static ChangeInfoResult TownHouseChangeInfo(uint hid, int numinfo, int prop, Byt
 				break;
 
 			case 0x17: // Four random colours to use
-				for (uint j = 0; j < 4; j++) housespec->random_colour[j] = buf->ReadByte();
+				for (uint j = 0; j < 4; j++) housespec->random_colour[j] = Colours(buf->ReadByte() & 0xF);
```
Comment (2TallTyler) [inline src/newgrf.cpp]: New magic number without explanation...?

File: src/company_gui.cpp (hunk 3 above)
Comment (PeterN) [reply src/company_gui.cpp:585]: It's a macro that does "PALETTE_RECOLOUR_START + (colour)", it is a bit unrelated but I changed it for consistency.

File: src/saveload/oldloader_sl.cpp (hunk 2 above)
Comment (PeterN) [reply src/saveload/oldloader_sl.cpp]: Hmm, probably but not relevant to this change I think.

File: src/newgrf.cpp (hunk 4 above)
Comment (PeterN) [reply src/newgrf.cpp]: Just a mask. I've switched it to `GB(x, 0, 4)` which is still fairly magic, but is at least used in other places so it's consistent.

Colours is a currently uint8_t but only values 0-15 are valid colours, and 255 is explicitly INVALID_COLOUR, which doesn't make sense to have here. Currently the code doesn't range check, which would end up treating a non-recolour sprite as a recolour sprite.

**Comment by PeterN** [comment]
Oops, while looking around I realised I missed `byte colour` on objects, so I'll tackle that later.

**Comment by PeterN** [comment]
> Casting is normally `(Colours)foo` rather than `Colours(foo)`

Now using C++ style explicit `static_cast<Colours>(foo)`

**Review by 2TallTyler — APPROVED** [review]

**Pull request merged**